src/transformers/models/layoutlmv3/modeling_tf_layoutlmv3.py

```diff
@@ -0,0 +1,1593 @@
+            head_mask = tf.expand_dims(head_mask, axis=-1)  # 1, 1, num_heads, 1, 1
+            head_mask = tf.tile(
+                head_mask, [self.config.num_hidden_layers, 1, 1, 1, 1]
+            )  # seq_length, 1, num_heads, 1, 1
+        elif n_dims == 2:
+            # Gets a tensor with masks for each layer (L) and head (H).
+            head_mask = tf.expand_dims(head_mask, axis=1)  # seq_length, 1, num_heads
+            head_mask = tf.expand_dims(head_mask, axis=-1)  # seq_length, 1, num_heads, 1
+            head_mask = tf.expand_dims(head_mask, axis=-1)  # seq_length, 1, num_heads, 1, 1
+        elif n_dims != 5:
+            raise ValueError(f"Wrong shape for head_mask (shape {head_mask.shape}).")
+        assert tf.rank(head_mask) == 5, f"Got head_mask rank of {tf.rank(head_mask)}, but require 5."
+        head_mask = tf.cast(head_mask, self.compute_dtype)
+        return head_mask
+
+    @unpack_inputs
+    def call(
+        self,
+        input_ids: Optional[tf.Tensor] = None,
+        bbox: Optional[tf.Tensor] = None,
+        attention_mask: Optional[tf.Tensor] = None,
+        token_type_ids: Optional[tf.Tensor] = None,
+        position_ids: Optional[tf.Tensor] = None,
+        head_mask: Optional[tf.Tensor] = None,
+        inputs_embeds: Optional[tf.Tensor] = None,
+        pixel_values: Optional[tf.Tensor] = None,
+        output_attentions: Optional[bool] = None,
+        output_hidden_states: Optional[bool] = None,
+        return_dict: Optional[bool] = None,
+        training: bool = False,
+    ) -> Union[
+        TFBaseModelOutput,
+        Tuple[tf.Tensor],
+        Tuple[tf.Tensor, tf.Tensor],
+        Tuple[tf.Tensor, tf.Tensor, tf.Tensor],
+    ]:
+        # This method can be called with a variety of modalities:
+        # 1. text + layout
+        # 2. text + layout + image
+        # 3. image
+        # The complexity of this method is mostly just due to handling of these different modalities.
+
+        output_attentions = output_attentions if output_attentions is not None else self.config.output_attentions
+        output_hidden_states = (
+            output_hidden_states if output_hidden_states is not None else self.config.output_hidden_states
+        )
+        return_dict = return_dict if return_dict is not None else self.config.return_dict
+
+        if input_ids is not None:
+            input_shape = tf.shape(input_ids)
+            batch_size = input_shape[0]
+            seq_length = input_shape[1]
+        elif inputs_embeds is not None:
+            input_shape = tf.shape(inputs_embeds)
+            batch_size = input_shape[0]
+            seq_length = input_shape[1]
+        elif pixel_values is not None:
+            batch_size = tf.shape(pixel_values)[0]
+        else:
+            raise ValueError("You have to specify either input_ids or inputs_embeds or pixel_values")
+
+        int_dtype = get_int_dtype(input_ids, attention_mask, token_type_ids, bbox)
```

Comment:
Not a fan of this function - could we maybe remove it and just put all the code here since it's only used once? It makes it quite hard to read the code (e.g. I now have to jump all the way up to understand this function)

Reply:
Sure 👍